### PR TITLE
init checkin to add LassoCV and RERF to optimizers

### PR DESCRIPTION
**Pull request by edcthayer**
Added LassoCrossValidated (LassoCV) and RegressionEnhancedRandomForest (RERF) regression models to the list of surrogate models available for optimizers.  This required creating MultiObjective versions for each of these regression models.  Fixed some bugs found via testing with random surrogate_model parameters.  Details below by file added/changed:

_source/Mlos.Python/mlos/Optimizers/BayesianOptimizerConfigStore.py_:

1. Added LassoCV, MultiObjectiveLassoCV, RERF, MultiObjective RERF to surrogate_model_implementation list and expanded the resulting hyper grid.  Didn't change the default from historical HomogeneousRandomForest.

_source/Mlos.Python/mlos/Optimizers/BayesianOptimizer.py_:

1. Added if-elif-else code to correctly instantiate the surrogate_model based on the optimizer_config.surrogate_model_implementation.
2. Extended assert test to include new surrogate models.

_source/Mlos.Python/mlos/Optimizers/RegressionModels/LassoCrossValidatedConfigStore.py_:
1. Corrected dimension type for LassoCV cv parameter (continuous --> discrete).
2. Restricted ranges on some model_configs to avoid Windows faults discovered in random config tests.

_source/Mlos.Python/mlos/Optimizers/RegressionModels/MultiObjectiveLassoCrossValidated.py_:
New class to allow LassoCV for multi-objective optimizations.

_source/Mlos.Python/mlos/Optimizers/RegressionModels/MultiObjectiveRegressionEnhancedRandomForest.py_:
New class to allow RERF for multi-objective optimizations.
Note: the .copy() on line 41 is needed b/c model_config.perform_initial_random_forest_hyper_parameter_search value is changed (True -> False) once grid search completes for the random forest fit().

_source/Mlos.Python/mlos/Optimizers/RegressionModels/MultiObjectiveRegressionEnhancedRandomForest.py_:
1. Correctly capture the random forest hyper parameters returned from the grid search (line 320).
2. Cleaned up some initializations.

_source/Mlos.Python/mlos/Optimizers/RegressionModels/unit_tests/TestMultiObjectiveLassoCrossValidated.py_:
New unit tests for new class.

_source/Mlos.Python/mlos/Optimizers/RegressionModels/unit_tests/TestMultiObjectiveRegressionEnhancedRandomForest.py_:
New unit tests for new class.

### CODE REVIEW

File: source/Mlos.Python/mlos/Optimizers/BayesianOptimizer.py [hunk 1]
```diff
@@ -59,20 +61,47 @@ def __init__(
 
         # Now let's put together the surrogate model.
         #
+        print(f'self.optimizer_config.surrogate_model_implementation: {self.optimizer_config.surrogate_model_implementation}')
```
Comment (byte-sculptor) [inline source/Mlos.Python/mlos/Optimizers/BayesianOptimizer.py:64]:
```suggestion
        self.logger.info(f'self.optimizer_config.surrogate_model_implementation: {self.optimizer_config.surrogate_model_implementation}')

```

File: source/Mlos.Python/mlos/Optimizers/BayesianOptimizer.py [hunk 2]
```diff
@@ -59,20 +61,47 @@ def __init__(
...
         # Note that even if the user requested a HomogeneousRandomForestRegressionModel, we still create a MultiObjectiveRegressionModel
         # with just a single RandomForest inside it. This means we have to maintain only a single interface.
         #
-        self.surrogate_model: MultiObjectiveRegressionModel = MultiObjectiveHomogeneousRandomForest(
-            model_config=self.optimizer_config.homogeneous_random_forest_regression_model_config,
-            input_space=self.optimization_problem.feature_space,
-            output_space=self.surrogate_model_output_space,
-            logger=self.logger
-        )
+        if self.optimizer_config.surrogate_model_implementation == HomogeneousRandomForestRegressionModel.__name__:
```
Comment (byte-sculptor) [inline source/Mlos.Python/mlos/Optimizers/BayesianOptimizer.py:75]: This is great! Now that we have so many regression models, I'll move this code to a RegressionModelFactory - it should make testing much easier :)

File: source/Mlos.Python/mlos/Optimizers/RegressionModels/LassoCrossValidatedConfigStore.py [hunk 3]
```diff
@@ -24,21 +24,19 @@ class Selection(Enum):
...
-            ContinuousDimension(name="eps", min=0, max=2 ** 16),
-            DiscreteDimension(name="num_alphas", min=0, max=10 ** 5),
+            ContinuousDimension(name="eps", min=0, max=10.0 ** -3),
+            DiscreteDimension(name="num_alphas", min=0, max=200),
             CategoricalDimension(name="fit_intercept", values=[False, True]),
             CategoricalDimension(name="normalize", values=[False, True]),
             CategoricalDimension(name="precompute", values=[False, True]),
-            DiscreteDimension(name="max_iter", min=0, max=10 ** 5),
-            ContinuousDimension(name="tol", min=0, max=2 ** 10),
+            DiscreteDimension(name="max_iter", min=100, max=5 * 10 **3),
```
Comment (byte-sculptor) [inline source/Mlos.Python/mlos/Optimizers/RegressionModels/LassoCrossValidatedConfigStore.py:32]:
```suggestion
            DiscreteDimension(name="max_iter", min=100, max=5 * (10 ** 3)),

```

File: source/Mlos.Python/mlos/Optimizers/RegressionModels/LassoCrossValidatedRegressionModel.py [hunk 4]
```diff
@@ -89,6 +91,10 @@ def __init__(
         self.partial_hat_matrix_ = 0
         self.regressor_standard_error_ = 0
 
+        # THE HACK
```
Comment (byte-sculptor) [inline source/Mlos.Python/mlos/Optimizers/RegressionModels/LassoCrossValidatedRegressionModel.py:94]: We may need to explain a little more here. If I remember right:

When LassoCV is used as part of RERF, it cannot reasonably compute the upper and lower bounds on its input space dimensions, as they are a polynomial combination of inputs to RERF. Thus, it approximates them with the empirical min and max. These approximations are biased: the lower bound is too large, the upper bound is too small. Consequently, during scoring, LassoCV is likely to see input outside of these bounds, but we still want LassoCV to produce predictions for those points. So we introduce a little hack: whenever LassoCV is instantiated as part of RERF, it should skip input filtering on predict. This field, controls this behavior.

Feel free to just copy-paste that in, or polish it to your liking!

File: source/Mlos.Python/mlos/Optimizers/RegressionModels/LassoCrossValidatedRegressionModel.py [hunk 5]
```diff
@@ -135,19 +141,28 @@ def fit(self, feature_values_pandas_frame, target_values_pandas_frame, iteration
...
-        if condition_number > 10.0 ** 10:
+        self.logger.info(
+            f'LassoCV: design_matrix condition number: {condition_number}'
+        )
+        if condition_number > 10.0 ** 4:
             # add small noise to x to remove singularity,
             #  expect prediction confidence to be reduced (wider intervals) by doing this
             self.logger.info(
-                f"Adding noise to design matrix used for prediction confidence due to condition number {condition_number} > 10^10."
+                f"Adding noise to design matrix used for prediction confidence due to condition number {condition_number} > 10^4."
```
Comment (byte-sculptor) [inline source/Mlos.Python/mlos/Optimizers/RegressionModels/LassoCrossValidatedRegressionModel.py:163]: 10**4
```suggestion
                f"Adding noise to design matrix used for prediction confidence due to condition number {condition_number} > 10**4."

```

Comment (byte-sculptor) [reply source/Mlos.Python/mlos/Optimizers/RegressionModels/LassoCrossValidatedRegressionModel.py:163]: It's clear what you mean... but my CDO strongly suggests that we should stick to the Python exponentiation operator :)

File: source/Mlos.Python/mlos/Optimizers/RegressionModels/MultiObjectiveLassoCrossValidated.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+#
+import logging
+from mlos.Optimizers.RegressionModels.LassoCrossValidatedConfigStore import lasso_cross_validated_config_store
+from mlos.Optimizers.RegressionModels.LassoCrossValidatedRegressionModel import LassoCrossValidatedRegressionModel
+from mlos.Optimizers.RegressionModels.NaiveMultiObjectiveRegressionModel import NaiveMultiObjectiveRegressionModel
+from mlos.Spaces import Hypergrid, Point, SimpleHypergrid
+
+
+class MultiObjectiveLassoCrossValidated(NaiveMultiObjectiveRegressionModel):
+    """Maintains multiple HomogeneousRandomForestRegressionModels each predicting a different objective.
```
Comment (byte-sculptor) [inline source/Mlos.Python/mlos/Optimizers/RegressionModels/MultiObjectiveLassoCrossValidated.py:13]: Lass
```suggestion
    """Maintains multiple LassoCVRegressionModels each predicting a different objective.

```

File: source/Mlos.Python/mlos/Optimizers/RegressionModels/MultiObjectiveLassoCrossValidated.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+            self,
+            model_type=LassoCrossValidatedRegressionModel,
+            model_config=model_config,
+            input_space=input_space,
+            output_space=output_space,
+            logger=logger
+        )
+
+
+        # We just need to assert that the model config belongs in homogeneous_random_forest_config_store.parameter_space.
```
Comment (byte-sculptor) [inline source/Mlos.Python/mlos/Optimizers/RegressionModels/MultiObjectiveLassoCrossValidated.py:35]:
```suggestion
        # We just need to assert that the model config belongs in lasso_cross_validated_config_store.parameter_space.

```

File: source/Mlos.Python/mlos/Optimizers/RegressionModels/MultiObjectiveRegressionEnhancedRandomForest.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+#
+import logging
+from mlos.Optimizers.RegressionModels.RegressionEnhancedRandomForestConfigStore import regression_enhanced_random_forest_config_store
+from mlos.Optimizers.RegressionModels.RegressionEnhancedRandomForestModel import RegressionEnhancedRandomForestRegressionModel
+from mlos.Optimizers.RegressionModels.NaiveMultiObjectiveRegressionModel import NaiveMultiObjectiveRegressionModel
+from mlos.Spaces import Hypergrid, Point, SimpleHypergrid
+
+
+class MultiObjectiveRegressionEnhancedRandomForest(NaiveMultiObjectiveRegressionModel):
+    """Maintains multiple HomogeneousRandomForestRegressionModels each predicting a different objective.
```
Comment (byte-sculptor) [inline source/Mlos.Python/mlos/Optimizers/RegressionModels/MultiObjectiveRegressionEnhancedRandomForest.py:13]:
```suggestion
    """Maintains multiple RegressionEnhancedRandomForestRegressionModel each predicting a different objective.

```

File: source/Mlos.Python/mlos/Optimizers/RegressionModels/MultiObjectiveRegressionEnhancedRandomForest.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+            self,
+            model_type=RegressionEnhancedRandomForestRegressionModel,
+            model_config=model_config,
+            input_space=input_space,
+            output_space=output_space,
+            logger=logger
+        )
+
+
+        # We just need to assert that the model config belongs in homogeneous_random_forest_config_store.parameter_space.
```
Comment (byte-sculptor) [inline source/Mlos.Python/mlos/Optimizers/RegressionModels/MultiObjectiveRegressionEnhancedRandomForest.py:35]:
```suggestion
        # We just need to assert that the model config belongs in regression_enhanced_random_forest_config_store.parameter_space.

```

File: source/Mlos.Python/mlos/Optimizers/RegressionModels/MultiObjectiveLassoCrossValidated.py [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+        # A more elaborate solution might be needed down the road, but for now this simple solution should suffice.
+        #
+        assert model_config in lasso_cross_validated_config_store.parameter_space
+
+        for output_dimension in output_space.dimensions:
+            print(f'output_dimension.name: {output_dimension.name}')
+            lasso_model = LassoCrossValidatedRegressionModel(
+                model_config=model_config,
```
Comment (byte-sculptor) [inline source/Mlos.Python/mlos/Optimizers/RegressionModels/MultiObjectiveLassoCrossValidated.py:43]: You copy the model_config in multi-objective RERF, but not here. Why?

File: source/Mlos.Python/mlos/Optimizers/RegressionModels/RegressionEnhancedRandomForestModel.py [hunk 11]
```diff
@@ -128,14 +133,23 @@ def trained(self) -> bool:
...
+
     def should_fit(
             self,
             num_samples: int
     ) -> bool:
-        root_base_model_should_fit = self.base_regressor_.should_fit(num_samples=num_samples)
-        # TODO : determine min sample needed to fit based on model configs
-        random_forest_should_fit = True
-        return root_base_model_should_fit and random_forest_should_fit
+        # since polynomial basis functions decrease the degrees of freedom (TODO: add reference),
```
Comment (byte-sculptor) [inline source/Mlos.Python/mlos/Optimizers/RegressionModels/RegressionEnhancedRandomForestModel.py:147]: This is neat :)

File: source/Mlos.Python/mlos/Optimizers/RegressionModels/unit_tests/TestMultiObjectiveLassoCrossValidated.py [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+
+        if objective_function_config_name == '2d_hypersphere_minimize_some':
+            num_training_samples = 25
+            num_testing_samples = 10
+        elif objective_function_config_name == '10d_hypersphere_minimize_some':
+            num_training_samples = 50
+            num_testing_samples = 10
+        elif objective_function_config_name == '5_mutually_exclusive_polynomials':
+            num_training_samples = 100
+            num_testing_samples = 50
```
Comment (byte-sculptor) [inline source/Mlos.Python/mlos/Optimizers/RegressionModels/unit_tests/TestMultiObjectiveLassoCrossValidated.py:43]:
```suggestion
            num_testing_samples = 50
        else:
            assert False

```

File: source/Mlos.Python/mlos/Optimizers/RegressionModels/unit_tests/TestMultiObjectiveRegressionEnhancedRandomForest.py [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+
+        if objective_function_config_name == '2d_hypersphere_minimize_some':
+            num_training_samples = 25
+            num_testing_samples = 10
+        elif objective_function_config_name == '10d_hypersphere_minimize_some':
+            num_training_samples = 50
+            num_testing_samples = 10
+        elif objective_function_config_name == '5_mutually_exclusive_polynomials':
+            num_training_samples = 100
+            num_testing_samples = 50
```
Comment (byte-sculptor) [inline source/Mlos.Python/mlos/Optimizers/RegressionModels/unit_tests/TestMultiObjectiveRegressionEnhancedRandomForest.py:42]:
```suggestion
            num_testing_samples = 50
        else:
            assert False

```

**Review by byte-sculptor — APPROVED** [review]

File: source/Mlos.Python/mlos/Optimizers/RegressionModels/MultiObjectiveLassoCrossValidated.py (hunk 10 above)
Comment (edcthayer) [reply source/Mlos.Python/mlos/Optimizers/RegressionModels/MultiObjectiveLassoCrossValidated.py:43]: Values in the model config are altered by the random forest GridSearchCV for the RERF.  When these configs are assigned to different objectives, they stomped all over each other.  I'll track down the lines in RERF model that alter the model_config and explain this in the MultiObjectiveRERF code where you've spotted this difference.

**Pull request merged**